### ISSUE TITLE
Key extraction fails when prefix causes byte repetition in msg

### ISSUE THREAD

**Issue by fasmat** [spacemeshos/go-spacemesh#3890]
## Description
When using a prefix, that causes the first few bytes of a message to repeat key extraction for ed25519 seems to fail.

## Steps to reproduce
I created a fuzzing test that allows us to reproduce this error: https://github.com/spacemeshos/go-spacemesh/blob/6ec1384b65052b59a9a2d5b674194f227136c38d/signing/keyExtractor_test.go

```golang
func Fuzz_PubKeyExtractor(f *testing.F) {
	f.Fuzz(func(t *testing.T, msg []byte, prefix []byte) {
		signer, err := NewEdSigner(WithPrefix(prefix))
		require.NoError(t, err)

		extractor, err := NewPubKeyExtractor(WithExtractorPrefix(prefix))
		require.NoError(t, err)

		sig := signer.Sign(msg)

		pub, err := extractor.Extract(msg, sig)
		require.NoError(t, err)
		require.Equal(t, pub.Bytes(), signer.PublicKey().Bytes())
	})
}
```

## Actual Behavior
If `prefix` repeats the first few bytes of `msg` the test above will fail pub key extraction. Here are a handful of examples where this happens:

| msg | prefix |
| ---- | ------ |
| `\x03\xe82x\x7f\xff` | `\x03\xe82` |
| `\\22x\x93o\x1d` | `\\22` |
| `---\x00\x00\x00` | `---` |
| `11,x\fx\f\xb4\xdf8\xb4\xdf8` | `11` |
| `11,x\f\xb4\xdf8` | `11` |

If the `prefix` does not cause repetition of bytes key extraction works without issues.

## Expected Behavior
Key extraction works with any prefix, even if it repeats the first few bytes of the msg.

## Environment
 - OS: Ubuntu 22.04
 - Node Version: develop
 - Smapp version (if applicable):  n/a

**Comment by fasmat** [comment]
@noamnelke this is a summary of the bug I mentioned in todays sync

**Comment by fasmat** [comment]
Here's an even simpler test without the abstractions of the `signing` package, but just using our `ed25519` implementation:

```golang
func Fuzz_PubKeyDirect(f *testing.F) {
	f.Fuzz(func(t *testing.T, msg []byte, prefix []byte) {
		pub, priv, _ := ed25519.GenerateKey(nil)
		sig := ed25519.Sign2(priv, append(prefix, msg...))
		pub2, _ := ed25519.ExtractPublicKey(append(prefix, msg...), sig)

		require.Equal(t, pub, pub2)
	})
}
```

**Comment by fasmat** [comment]
After some additional investigation I found the issue: the reason this happens is because `msg` and `prefix` are re-sliced from the same underlying array during a fuzzing test.

The test fails if `msg` and `prefix` point at the same location of this array and `prefix` is shorter than `msg`. The `append` will then change the contents of `msg`. Here is a link that shows how this happens: https://go.dev/play/p/HepPPj0HNoh

Although this is unlikely to happen outside of fuzzing, I updated the code to make sure this can't happen. The fix was to create a new byte-slice and `copy` the data instead of using `append`.
